keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1251 @@
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Future Improvements
+
+### Disruption Controller: Eviction Protection and Observability
+
+With the `.spec.MinAvailable` and `.spec.MaxUnavailable` options, PDB could unfortunately allow
+surge pods to be evicted because `.status.CurrentHealthy` could become higher than
+`.status.desiredHealthy`. This could disrupt ongoing migration of an application. To support
+mission-critical applications, we might do one of the following:
+- Consider pods with an `EvacuationInitiated` condition as disrupted and thus decreasing
+  `.status.CurrentHealthy`.
+- Or increase the `.status.desiredHealthy` by the number of pods with the `EvacuationInitiated`
+  condition. This would keep the `.status.DisruptionsAllowed` value low enough not to disrupt the
+  migration.
+
+We can also count the evacuating pods in the status for observability, as this feature changes the
+behavior of the PodDisruptionBudget status.
+
+
+```golang
+type PodDisruptionBudgetStatus struct {
+    ...
+    // total number of pods that are evacuating and expected to be terminated
+    EvacuatingPods int32
+    ...
+}
+```
+
```

Comment:
A future improvement could be to set a boolean condition on a workload management object (StatefulSet, ReplicaSet, Job etc) if any Pod managed by that object is matched by any non-expired NodeMaintenance.

_This feedback should not block merging the KEP as provisional._

Reply:
mentioning this in https://github.com/kubernetes/enhancements/pull/4565 in the `Follow-up Design Details for Kubernetes Workloads` section